### ISSUE TITLE
Custom Loss Function for LGBMRegressor

### ISSUE THREAD

**Issue by jdtrebbien** [microsoft/LightGBM#5256]
I want to use a custom loss function for LGBMRegressor but I cant find any documentation on it. If I understand it correctly I need to use the params 'objective' and 'metric' to completely change the loss function in training and evaluation. The function for 'objective' returning (grad, hess) and the function for 'metric' returning ('<loss_name>', loss, uses_max). I am just searching for the two functions that are being used when the default objective 'regression' (l2 loss) is beeing used so I can reproduce and change it. I already found the C++ code for the regression, but I am unable to reproduce it using two custom functions written in python. 

This would be my approach:

```
def custom_l2_loss(y_pred, y_train):
    ...
    return grad, hess

def custom_l2_eval(y, data):
    ...
    return 'l2', loss.mean(), False


model = LGBMRegressor(
    objective=custom_l2_loss,
    metric=custom_l2_eval
)

model.fit(
    X_train,
    y_train,
    eval_set=[
        (X_val, y_val),
    ]
)
```

Does someone know how to reproduce the l2_loss in python?

**Comment by jmoralez** [comment]
Hi @jdtrebbien, thank you for your interest in LightGBM. If you only want a custom objective it is enough to specify `objective`, you can find an example of l2 here: https://github.com/microsoft/LightGBM/blob/4971a06668df7eabeb7d4bb1987abb442f2970c9/tests/python_package_test/test_sklearn.py#L207-L214
Keep in mind that the results will be a bit different at first because of the different init score https://github.com/microsoft/LightGBM/issues/5114#issuecomment-1084994020 but if you train for enough iterations you should get the same results.
You can also use the built-in metrics if you want, there's no need to use a custom metric if there's already a built-in one.

Please let us know if this helps.

**Comment by jdtrebbien** [comment]
Thank you very much, that is basically what I was looking for.  But are you sure I dont need to change metric or something to also change the loss function being used on the validation set for early_stopping? Since it still says `valid_0's l2: <num>` for the validation set and not the name of my custom function.

One more thing if you have the time:
When using the custom l2 loss from this, any idea how to get the l1/2 loss? That is what I need to implement since its not one of the built-in metrics. For the normal loss it would be `(sum((x - y) ** 0.5) ** 2)` instead of the normal l2 loss with `(sum((x - y) ** 2) ** 0.5)` but I am a bit stuck when it comes to defining this for grad and hess...

**Comment by jmoralez** [comment]
That's the default metric for regression, if you do want to use a custom metric as well you have to set `metric='None'` and pass your custom metric on the fit method like here:
https://github.com/microsoft/LightGBM/blob/4971a06668df7eabeb7d4bb1987abb442f2970c9/tests/python_package_test/test_sklearn.py#L913-L917

TBH I don't know about the l1/2 loss, the square root is defined only for non-negative numbers so I don't think it'll be a good objective but maybe someone else here can comment on it.

**Comment by StatMixedML** [comment]
@jdtrebbien You can find an end-to-end way of how to use custom-loss and evaluation function on my [LightGBMLSS Repo](https://github.com/StatMixedML/LightGBMLSS/blob/master/lightgbmlss/distributions/Gaussian_AutoGrad.py#L92)

For the linked example, I use PyTorch's autograd function, so that you can get gradients and hessians for any user-defined loss.

Let me know if that is useful.
